metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/storage/BlockPruning.scala

```diff
@@ -0,0 +1,37 @@
+package io.iohk.metronome.hotstuff.service.storage
+
+import io.iohk.metronome.hotstuff.consensus.basic.Agreement
+import io.iohk.metronome.storage.KVStore
+
+import cats.implicits._
+
+object BlockPruning {
+
+  /** Prune blocks which are not descendants of the N-th ancestor of the last executed block. */
+  def prune[N, A <: Agreement](
+      blockStorage: BlockStorage[N, A],
+      viewStateStorage: ViewStateStorage[N, A],
+      blockHistorySize: Int
+  ): KVStore[N, Unit] = {
+    for {
+      // Always keep the last executed block.
+      lastExecutedBlock <- viewStateStorage.getLastExecutedBlockHash.lift
+      pathFromRoot      <- blockStorage.getPathFromRoot(lastExecutedBlock).lift
+
+      // Everything but the last N blocks in the chain leading up to the
+      // last executed block can be pruned.
+      pruneable = pathFromRoot.dropRight(blockHistorySize)
```

Comment:
Ok, I guess the double reversal confused me. But also the fact that with this we are actually keeping `N + 1` blocks, aren't we?

What about something like:
```maybeNewRoot = pathFromRoot.reverse.lift(blockHistorySize - 1)```

Reply:
Yeah, it's N+1, I thought that was close enough :)

Yours is a nice solution too. I thought having this interim `pruneable` variable is a good mental anchor. But we can go with your version.

Reply:
Done.